src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,194 @@
+                CreateProjectInstance(projectCollection),
+                targetsToBuild: ["Build"]);
+            var buildResult = BuildManager.DefaultBuildManager.BuildRequest(buildRequest);
+            if (buildResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            BuildManager.DefaultBuildManager.EndBuild();
+            return 0;
+        }
+        catch (Exception e)
+        {
+            Console.Error.WriteLine(e.Message);
+            return 1;
+        }
+        finally
+        {
+            foreach (var (key, value) in savedEnvironmentVariables)
+            {
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            foreach (var binaryLogger in binaryLoggers)
+            {
+                binaryLogger.Shutdown();
+            }
+
+            consoleLogger.Shutdown();
+        }
+
+        static ImmutableArray<ILogger> GetBinaryLoggers(string[] args)
+        {
+            return args
+                .Where(RunCommand.IsBinLogArgument)
+                .Select(static ILogger (arg) => new BinaryLogger
+                {
+                    Parameters = arg.IndexOf(':') is >= 0 and var index
+                            ? arg[(index + 1)..]
+                            : "msbuild.binlog",
+                })
+                .ToImmutableArray();
+        }
+    }
+
+    public ProjectInstance CreateProjectInstance(ProjectCollection projectCollection)
+    {
+        return CreateProjectInstance(projectCollection, addGlobalProperties: null);
+    }
+
+    private ProjectInstance CreateProjectInstance(
+        ProjectCollection projectCollection,
+        Action<IDictionary<string, string>>? addGlobalProperties)
+    {
+        var projectRoot = CreateProjectRootElement(projectCollection);
+
+        var globalProperties = projectCollection.GlobalProperties;
+        if (addGlobalProperties is not null)
+        {
+            globalProperties = new Dictionary<string, string>(projectCollection.GlobalProperties, StringComparer.OrdinalIgnoreCase);
+            addGlobalProperties(globalProperties);
+        }
+
+        return ProjectInstance.FromProjectRootElement(projectRoot, new ProjectOptions
+        {
+            GlobalProperties = globalProperties,
+        });
+    }
+
+    private ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+    {
+        var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+        var projectFileText = """
+            <Project>
+                <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+
+                <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net9.0</TargetFramework>
+                    <ImplicitUsings>enable</ImplicitUsings>
+                    <Nullable>enable</Nullable>
+                </PropertyGroup>
```

Comment:
Should we enable `AllowUnsafeBlocks` by default?

Reply:
Think we should keep the current state where unsafe requires a specific build opt in. Overall we want to have less unsafe code in the ecosystem, not more.

Reply:
In general, the `dotnet run file.cs` feature aims to use the *default* project template, whatever that is. That makes it predictable.

Reply:
If that's true, then why did you change the default template?

Reply:
I made minimal changes to workaround bugs that prevent it building without a physical file on disk. There is a tracking issue linked and ideally that will be fixed and the workaround can be removed and the template will exactly match the default one.